File: Divide-and-Conquer Searching/Binary Search/Recursive Binary Search/RecursiveBinarySearch.py
def recursive_binary_search(arr, left, right, target):
    if left > right:
        return -1

    mid = left + (right - left) // 2

    if arr[mid] == target:
        return mid

    if arr[mid] < target:
        return recursive_binary_search(
            arr,
            mid + 1,
            right,
            target
        )

    return recursive_binary_search(
        arr,
        left,
        mid - 1,
        target
    )


def main():
    arr = [5, 12, 18, 25, 31, 42, 56]
    target = 42

    index = recursive_binary_search(
        arr,
        0,
        len(arr) - 1,
        target
    )

    if index != -1:
        print(
            f"Element {target} found at index {index}"
        )
    else:
        print(
            f"Element {target} not found"
        )


if __name__ == "__main__":
    main()

File: Divide-and-Conquer Searching/Binary Search/Recursive Binary Search/RecursiveBinarySearch.java
public class RecursiveBinarySearch {

    public static int recursiveBinarySearch(
            int[] arr,
            int left,
            int right,
            int target) {

        if (left > right) {
            return -1;
        }

        int mid = left + (right - left) / 2;

        if (arr[mid] == target) {
            return mid;
        }

        if (arr[mid] < target) {
            return recursiveBinarySearch(
                    arr,
                    mid + 1,
                    right,
                    target
            );
        }

        return recursiveBinarySearch(
                arr,
                left,
                mid - 1,
                target
        );
    }

    public static void main(String[] args) {

        int[] arr = {
                5, 12, 18, 25, 31, 42, 56
        };

        int target = 42;

        int index = recursiveBinarySearch(
                arr,
                0,
                arr.length - 1,
                target
        );

        if (index != -1) {
            System.out.println(
                    "Element " + target +
                    " found at index " + index
            );
        } else {
            System.out.println(
                    "Element " + target +
                    " not found"
            );
        }
    }
}
